option for disabling cheats

Laserlicht opened the pull request:
Until we have a better place, it's in turn options:
![grafik](https://github.com/vcmi/vcmi/assets/13953785/de610640-e5ef-47f3-a25a-35a3d5639719)

@IvanSavenko We should add this to develop and not beta because it breaks savegame, right?

dydzio0614 commented:
i think this is not good GUI place for this option

IvanSavenko commented:
yeah, probably for develop.
However please bump SERIALIZATION_VERSION constant in CSerializer.h and add version check for your new flag in startinfo.

This way saves from 1.4 can still be loaded in 1.4 (but multiplayer won't be possible)

IvanSavenko commented:
> i think this is not good GUI place for this option

True, but can't think of any other place for this option at the moment.

IvanSavenko commented on client/lobby/OptionsTabBase.cpp:
```diff
@@ -90,6 +90,10 @@ OptionsTabBase::OptionsTabBase(const JsonPath & configPath)
 		CSH->setSimturnsInfo(info);
 	});
 
+	addCallback("setCheatAllowed", [&](int index){
+		CSH->setCheatAllowedInfo(index);
+	});
+
```
You also need to set this checkbox to correct value in recreate() method

IvanSavenko commented on lib/registerTypes/RegisterTypesLobbyPacks.h:
```diff
@@ -54,6 +54,7 @@ void registerTypesLobbyPacks(Serializer &s)
 	s.template registerType<CLobbyPackToServer, LobbySetPlayer>();
 	s.template registerType<CLobbyPackToServer, LobbySetPlayerName>();
 	s.template registerType<CLobbyPackToServer, LobbySetTurnTime>();
+	s.template registerType<CLobbyPackToServer, LobbySetCheatAllowed>();
```
please add to the very end (even if it won't be alphabetical)

IvanSavenko replied:
And if you are curious as to "why":
registerTypesXXX methods generate enumeration of types that can be (de)serialized.

Each type is serialized as (numeric type ID) (serialized type data)
So changing order will lead to a crash if game attempts to load type with wrong type ID

Laserlicht commented:
> yeah, probably for develop. However please bump SERIALIZATION_VERSION constant in CSerializer.h and add version check for your new flag in startinfo.

Not sure how to check exactly. Haven't found any examples.

IvanSavenko commented:
> Not sure how to check exactly. Haven't found any examples.

```
if (version >= new value )
    h & newField;
else
    h = defaultValue;
```

dydzio0614 commented:
well, another minor doubts I have:

1. before this feature ai simturns option is isolated and harder to enable by mistake this way
2. there will probably be some simturn behavior flip switch which would be fitting to add somewhere to this screen

Laserlicht commented:
> well, another minor doubts I have:
> 
>     1. before this feature ai simturns option is isolated and harder to enable by mistake this way
> 
>     2. there will probably be some simturn behavior flip switch which would be fitting to add somewhere to this screen

Yes basically we should have an "other options" screen similar to turn options. But we only have this option yet...

Laserlicht commented:
Added extra page for additional options:
![grafik](https://github.com/vcmi/vcmi/assets/13953785/b4edcb0c-9d6c-4525-9b17-acbe867bf1a1)

Much space for additional improvements. Moved the turn options up.

Also added option for unlimited battle replay (so we have two settings yet).

Missing is a possibility to open this window within bonus selection. Maybe I add this in separate PR later.

IvanSavenko commented:
I would like to ask our community manager @dydzio0614 whether there were other similar requests before, since this give is more space for potential game options

dydzio0614 commented:
One doubt i have is where we will add button for banning/bidding panel

Another doubt is about what should be serialized or allowed to be changed on the fly the way gameconfig is changeable right now - i think most player requests  would rather fit being parts of gameconfig changeable - hota dimension door mechanics, hota diplomacy nerf, allowing glitch for heroes without army, fixing captain hat unlimited movement points the hota way...

The thing that we need separately anyway is saving last used settings for random map generator, turn options and this menu

Laserlicht commented:
> One doubt i have is where we will add button for banning/bidding panel

As singleplayer guy: What do you mean? Something like the PvP options from HD mod? I think we can combine it with the Chat button (makes most sense to me). Maybe this one button as threestate switch (Scenario info; Chat; PvP). Or a sub button if chat is enabled.

> Another doubt is about what should be serialized or allowed to be changed on the fly the way gameconfig is changeable right now - i think most player requests would rather fit being parts of gameconfig that is changeable on save reload - hota dimension door mechanics, hota diplomacy nerf, allowing glitch for heroes without army, fixing captain hat unlimited movement points the hota way...

Yeah, but I think some options should be have to predefined and shouldn't be changeable ingame. Especially with multiplayer where the host can set this settings and the player shouldn't able to change it.

dydzio0614 commented:
yeah i mean pvp options from hd mod

dydzio0614 commented:
and i cannot think of anything else than these two options that we have for now - things I listed are probably 2x more important than stuff I will be able to invent for extra options tab...

Laserlicht commented:
> and i cannot think of anything else than these two options that we have for now - things I listed are probably 2x more important than stuff I will be able to invent for extra options tab...

Yeah, but these two (or one) options are not fitting anywhere else. That was the main problem here. HD uses a popup here for additional options. But I think a page is better.

dydzio0614 commented:
yeah, let it be 2 options for now

Laserlicht commented:
@IvanSavenko Okay from your side?

IvanSavenko commented:
Not against it I guess, although have same questions as Dydzio, specifically - that with this we won't have any free space in that screen.

Let me check code once more, but probably looks OK
